packages/oidc-auth/src/index.ts

```diff
@@ -0,0 +1,289 @@
+    authorizationRequestUrl.searchParams.set('scope', 'openid email offline_access')
+  }
+  authorizationRequestUrl.searchParams.set('state', state)
+  authorizationRequestUrl.searchParams.set('nonce', nonce)
+  authorizationRequestUrl.searchParams.set('code_challenge', code_challenge)
+  authorizationRequestUrl.searchParams.set('code_challenge_method', 'S256')
+  if (as.issuer === 'https://accounts.google.com') {
+    // Google requires 'access_type=offline' and 'prompt=consent' to obtain a refresh token
+    authorizationRequestUrl.searchParams.set('access_type', 'offline')
+    authorizationRequestUrl.searchParams.set('prompt', 'consent')
+  }
+  return authorizationRequestUrl.toString()
+}
+
+/**
+ * Processes the OAuth2 callback request.
+ */
+export const processOAuthCallback = async (c: Context) => {
+  const as = await getAuthorizationServer(c)
+  const client = getClient(c)
+
+  // Parses the authorization response and validates the state parameter
+  const state = getCookie(c, 'state')
+  deleteCookie(c, 'state')
+  const currentUrl: URL = new URL(c.req.url)
+  const params = oauth2.validateAuthResponse(as, client, currentUrl, state)
+  if (oauth2.isOAuth2Error(params)) {
+    throw new Error(`OAuth2Error: [${params.error}] ${params.error_description}`)
+  }
+
+  // Exchanges the authorization code for a refresh token
+  const code = c.req.query('code')
+  const nonce = getCookie(c, 'nonce')
+  deleteCookie(c, 'nonce')
+  const code_verifier = getCookie(c, 'code_verifier')
+  deleteCookie(c, 'code_verifier')
+  const continue_url = getCookie(c, 'continue')
+  deleteCookie(c, 'continue')
+  if (code === undefined || nonce === undefined || code_verifier === undefined) {
+    throw new Error('Missing required parameters / cookies')
+  }
+  const result = await exchangeAuthorizationCode(as, client, params, c.env.OIDC_REDIRECT_URI, nonce, code_verifier)
+  await setAuth(c, result)
+  return c.redirect(continue_url || '/')
+}
+
+/**
+ * Exchanges the authorization code for a refresh token.
+ */
+const exchangeAuthorizationCode = async (as: oauth2.AuthorizationServer, client: oauth2.Client, params: URLSearchParams, redirect_uri: string, nonce: string, code_verifier: string) => {
+  const response = await oauth2.authorizationCodeGrantRequest(
+    as,
+    client,
+    params,
+    redirect_uri,
+    code_verifier,
+  )
+  // Handle www-authenticate challenges
+  const challenges = oauth2.parseWwwAuthenticateChallenges(response)
+  if (challenges !== undefined) {
+    for (const challenge of challenges) {
+      console.log('challenge: ', challenge)
+    }
+    throw new Error()
+  }
+  const result = await oauth2.processAuthorizationCodeOpenIDResponse(as, client, response, nonce)
+  if (oauth2.isOAuth2Error(result)) {
+    throw new Error(`OAuth2Error: [${result.error}] ${result.error_description}`)
+  }
+  return result
+}
+
+/**
+ * Returns a middleware that requires OIDC authentication.
+ */
+export const oidcAuthMiddleware = ()  => {
```

Comment:
I think it would be better to pass the variables for this middleware as arguments instead of environment variables. Like the following:

```ts
app.use('*', oidcAuthMiddleware({
  sessionSecret: 'xxx',
  sessionRefreshInterval: 123,
  sessionExpires: 123,
  ...
}))
```

Reply:
I understand that passing configuration via environment variables is not a good idea. It should have portability issues.

However, if settings are passed as constructor arguments for middleware, the settings cannot be used in paths where middleware is not applied.

Authentication is required for paths to which `oidcAuthMiddleware` is applied. On the other hand, the callback and logout path should be processed without authentication, so we cannot apply `oidcAuthMiddleware` to these paths. Nevertheless, `revokeSession` and `processOAuthCallback` functions used in these paths require OIDC endpoint information.

For the above reasons, I created a separate function for configuration.

Reply:
Thank you for the explanation! Understood well.

And I know you created the `configureOidcAuth,` which seems to be good, but I think there is a problem.

For Cloudflare Workers, we can't get the environment variables globally. We can get only in the handler. So, we have to use the approach such as Bearer Auth middleware does like the following:

```ts
import { bearerAuth } from 'hono/bearer-auth'

type Bindings = {
  TOKEN: string
}

const app = new Hono<{ Bindings: Bindings }>()

app.get('/auth/*', async (c, next) => {
  const auth = bearerAuth({
    token: c.env.TOKEN,
  })
  await auth(c, next)
})
```

Based on that, I see two ways to go about it.

### 1. Options pattern

In the case of Cloudflare, we want to use environments, but as you mentioned, they are used in other functions. The solution is to have a function like `setupOidcAuthMiddleware` and write it at the top to get environment variables first.

```ts
const app = new Hono()

// Put this at the top
app.use('*', setupOidcAuthMiddleware({
  sessionSecret: 'xxx',
  sessionRefreshInterval: 123,
  sessionExpires: 123,
  ...
}))
```

### 2. Environment variables pattern

The other is to get the value of the setting in an environment variable, as you did at first. For this, `env()` in `hono/adapter` may be useful. This will give you multi-runtime support.

https://hono.dev/helpers/adapter#env

---

Either way is fine, but 2 might be better since it is opnionative. Also, it might be good to see how other middleware solves that problem.

Reply:
Thank you for your detailed feedback on my code.

I have followed your 2nd suggestion, making use of environment variables for the middleware's configuration, and have successfully implemented it using `env()` from `hono/adapter`. This indeed provides us with the flexibility of multi-runtime support, which I find quite beneficial. For example, we can use it on Netlify to provide OpenID authentication to static pages.

I've committed these changes and pushed them to the repository. Could you please review them when you get a chance?

Thank you again for your constructive advice.

Reply:
I think that choice is good!

We will release version 4.0.0 tomorrow, so let's hold off on the merge until then. I may have the `peerDependencies` changed.